### PR TITLE
Add support for packaging analyzers into ref-pack

### PR DESCRIPTION
**Pull request by ericstj**
Related: https://github.com/dotnet/runtime/issues/54687
Spec: https://github.com/dotnet/designs/blob/main/accepted/2021/InboxSourceGenerators.md

We'll start flowing source-generators through the transport package in https://github.com/dotnet/runtime/pull/54950.  This will pick them up and include them in the ASP.NET refpack.

The framework list change is a port of https://github.com/dotnet/arcade/pull/7561.

### CODE REVIEW

File: eng/tools/RepoTasks/CreateFrameworkListFile.cs [hunk 1]
```diff
@@ -57,12 +57,34 @@ public override bool Execute()
...
+                    }
+
+                    var pathParts = path.Split('/');
+
+                    if (pathParts.Length < 3 || !pathParts[1].Equals("dotnet", StringComparison.Ordinal) || pathParts.Length > 4)
+                    {
+                        Log.LogError($"Unexpected analyzer path format {path}.  Expected  'analyzers/dotnet(/language)/analyzer.dll");
+                    }
+
+                    if (pathParts.Length > 3)
```
Comment (dougbu) [inline eng/tools/RepoTasks/CreateFrameworkListFile.cs:81]: Should this be
```suggestion
                    if (pathParts.Length >= 3)
```

File: eng/tools/RepoTasks/CreateFrameworkListFile.cs [hunk 2]
```diff
@@ -57,12 +57,34 @@ public override bool Execute()
...
-                    "File",
-                    new XAttribute("Type", f.IsNative ? "Native" : "Managed"),
-                    new XAttribute(
-                        "Path",
-                        Path.Combine(f.PackagePath, f.Filename).Replace('\\', '/')));
+                string path = Path.Combine(f.PackagePath, f.Filename).Replace('\\', '/');
+                string type = f.IsNative ? "Native" : "Managed";
+                var element = new XElement("File", new XAttribute("Path", path));
+
+                if (path.StartsWith("analyzers/"))
```
Comment (dougbu) [inline eng/tools/RepoTasks/CreateFrameworkListFile.cs:64]:
```suggestion
                if (path.StartsWith("analyzers/", StringComparison.Ordinal))
```

File: eng/tools/RepoTasks/CreateFrameworkListFile.cs [hunk 3]
```diff
@@ -57,12 +57,34 @@ public override bool Execute()
...
+                {
+                    type = "Analyzer";
+
+                    if (path.EndsWith(".resources.dll"))
```
Comment (dougbu) [inline eng/tools/RepoTasks/CreateFrameworkListFile.cs:68]:
```suggestion
                    if (path.EndsWith(".resources.dll", StringComparison.Ordinal))
```

File: eng/tools/RepoTasks/CreateFrameworkListFile.cs (hunk 1 above)
Comment (ericstj) [reply eng/tools/RepoTasks/CreateFrameworkListFile.cs:81]: I specifically want it to be greater than 3.  That means there is a language directory.  If it's only 3 there is no language directory.

Comment (ericstj) [reply eng/tools/RepoTasks/CreateFrameworkListFile.cs:81]:
```suggestion
                    // Check if we have enough parts for language directory and include it
                    if (pathParts.Length > 3)
```

File: eng/tools/RepoTasks/CreateFrameworkListFile.cs [hunk 4]
```diff
@@ -57,12 +57,34 @@ public override bool Execute()
...
+                    var pathParts = path.Split('/');
+
+                    if (pathParts.Length < 3 || !pathParts[1].Equals("dotnet", StringComparison.Ordinal) || pathParts.Length > 4)
+                    {
+                        Log.LogError($"Unexpected analyzer path format {path}.  Expected  'analyzers/dotnet(/language)/analyzer.dll");
+                    }
+
+                    if (pathParts.Length > 3)
+                    {
+                        element.Add(new XAttribute("Language", pathParts[2]));
```
Comment (dougbu) [inline eng/tools/RepoTasks/CreateFrameworkListFile.cs:84]: > I specifically want it to be greater than 3. That means there is a language directory. If it's only 3 there is no language directory.

In that case, this line should be
```suggestion
                        element.Add(new XAttribute("Language", pathParts[3]));
```

Comment (ericstj) [reply eng/tools/RepoTasks/CreateFrameworkListFile.cs:84]: No, that would be using the file name as language.  Length is 1-base, array indexes are 0-base.  When we have more than 3 parts we want to take the 3rd part (index 2).

Comment (ericstj) [reply eng/tools/RepoTasks/CreateFrameworkListFile.cs:84]: I added a comment, please see if that clarifies it.

Comment (dougbu) [reply eng/tools/RepoTasks/CreateFrameworkListFile.cs:84]: Yeah. I just somehow forgot the filename portion of the path 🤦

**Comment by dougbu** [comment]
Separately, will the CI actually validate the new code path prior to dotnet/runtime#54950 getting in and a new runtime build flowing into this repo❔

**Comment by ericstj** [comment]
> Separately, will the CI actually validate the new code path prior to dotnet/runtime#54950 getting in and a new runtime build flowing into this repo❔

Not at the moment (it's a feature that lets the work remain decoupled :)).  Do you have a place where you test that FrameworkList tasks?

File: eng/tools/RepoTasks/CreateFrameworkListFile.cs (hunk 4 above)
Comment (ericstj) [reply eng/tools/RepoTasks/CreateFrameworkListFile.cs:84]: It's ok, you were spot on for catching an off-by-one error which this does appear to be.  Better to add a comment.

**Comment by dougbu** [comment]
> Do you have a place where you test that FrameworkList task?

Yes, please see https://github.com/dotnet/aspnetcore/blob/main/src/Framework/test/TargetingPackTests.cs#L297-L393

**Comment by ericstj** [comment]
Those tests appear to be testing the product content, not running the task itself with test-inputs, is that correct?

In either case I should update those tests so that they don't break when the actual transport package arrives.  Is there any other place where you'd want this to be tested in lieu of the transport package?

**Comment by dougbu** [comment]
> Those tests appear to be testing the product content, not running the task itself with test-inputs, is that correct?

Yes. Our `RepoTasks` project has no associated unit tests, partially because later stages of the build exercise most of the tasks pretty completely. We're adding an untested code path here though the same code may already be exercised elsewhere.

> In either case I should update those tests so that they don't break when the actual transport package arrives. Is there any other place where you'd want this to be tested in lieu of the transport package?

No that seems sufficient. I like the thought of not only being ready in advance of whatever Maestro++ PR brings in your  dotnet/runtime#54950 work but also doing more than just ignoring the new files listed in the framework list.

**Comment by dougbu** [comment]
I don't see any failures clearly tied to your changes but would rather merge on green.

Most test failures look like transient problems but the `Reset_PriorOSVersions_NotSupported()` failure on Windows is a known issue (see #34003). Fix (see #34004) should get in in about an hour if other flakiness doesn't hurt the CI checks. Then you'll be able to `/azp run` to get this validating (also hopefully successfully).

**Comment by dougbu** [comment]
Separately I saw you merged dotnet/runtime#54950 and fix should have affected Microsoft.AspNetCore.Internal.Transport and Microsoft.NETCore.App.Ref v6.0.0-preview.7.21351.5 (from build https://dev.azure.com/dnceng/internal/_build/results?buildId=1215933&view=results) or newer. However, while System.Text.Json.SourceGeneration.dll was in Microsoft.NETCore.App.Ref, Microsoft.Extensions.Logging.Generators.dll was _not_ in Microsoft.AspNetCore.Internal.Transport. The transport package still lacks an analyzers/ folder. Is something more needed before that shows up❔

**Comment by ericstj** [comment]
> The transport package still lacks an analyzers/ folder. Is something more needed before that shows up❔

Good call.  Let me fix that.  My test used a manual mockup on the version ASP.NET is using today.

**Comment by ericstj** [comment]
Seems to be passing now @dougbu.  Also fix should be in here: https://github.com/dotnet/runtime/pull/55042

**Review by dougbu — APPROVED** [review]
Should have done this hours ago :shipit:

**Pull request merged**